pkg/metrics/collector/sockstat.go

```diff
@@ -0,0 +1,165 @@
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package collector
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type sockStatCollector struct {
+	inuse  *prometheus.Desc
+	tw     *prometheus.Desc
+	memory *prometheus.Desc
+}
+
+var pageSize = os.Getpagesize()
+
+func init() {
+	registerCollector("socket", NewSockStatCollector)
+}
+
+// NewSockStatCollector returns a new Collector exposing socket stats.
+func NewSockStatCollector() (Collector, error) {
+	return &sockStatCollector{
+		inuse: prometheus.NewDesc(
+			"num_inuse_sockets",
+			"Number of inuse sockets",
+			[]string{"protocol"}, nil,
+		),
+		tw: prometheus.NewDesc(
+			"num_tw_sockets",
+			"Number of sockets in time wait state",
+			nil, nil,
+		),
+		memory: prometheus.NewDesc(
+			"socket_memory",
+			"Amount of memory used by sockets",
+			nil, nil,
+		),
+	}, nil
+}
+
+type socketStats struct {
+	tcpInuse uint64 // num of tcp inuse sockets
+	udpInuse uint64 // num of udp inuse sockets
+	tcpTw    uint64 // num of tcp tw sockets
+	mem      uint64 // sum of tcp and udp memory
+}
+
+func getSockStats(fileName string) (*socketStats, error) {
+	file, err := os.Open(fileName)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	return parseSockStats(file)
+}
+
+func parseSockStats(r io.Reader) (*socketStats, error) {
+	var ss socketStats
+
+	scanner := bufio.NewScanner(r)
+	for scanner.Scan() {
+		in := scanner.Text()
+		line := strings.Split(in, " ")
+		if len(line) < 2 || len(line[0]) < 2 || line[0][len(line[0])-1] != ':' {
+			return nil, fmt.Errorf("invalid input %q", in)
+		}
+		// Remove trailing ':'.
+		protocol := line[0][:len(line[0])-1]
+		if protocol == "TCP" || protocol == "TCP6" {
+			for i := 1; i < len(line) && i+1 < len(line); i += 2 {
+				switch line[i] {
+				case "inuse":
+					v, err := strconv.ParseUint(line[i+1], 10, 64)
+					if err != nil {
+						return nil, fmt.Errorf("invalid value %q in sockstats: %v", line[i+1], err)
+					} else {
+						ss.tcpInuse = v
+					}
+				case "tw":
+					v, err := strconv.ParseUint(line[i+1], 10, 64)
+					if err != nil {
+						return nil, fmt.Errorf("invalid value %q in sockstats: %v", line[i+1], err)
+					} else {
+						ss.tcpTw = v
+					}
+				case "mem":
+					v, err := strconv.ParseUint(line[i+1], 10, 64)
+					if err != nil {
+						return nil, fmt.Errorf("invalid value %q in sockstats: %v", line[i+1], err)
+					} else {
+						ss.mem += v
```

Comment:
why plus here vs set in the other places?

Reply:
Because this mem is the sum of tcp mem and udp mem. The others are not sum across tcp and udp.

Reply:
Added the comment.